pkg/fuzzer/fuzzer.go

```diff
@@ -0,0 +1,307 @@
+	"github.com/google/syzkaller/prog"
+)
+
+type Fuzzer struct {
+	Config    *Config
+	Corpus    *Corpus
+	NewInputs chan rpctype.Input
+
+	ctx    context.Context
+	mu     sync.Mutex
+	stats  map[string]uint64
+	rnd    *rand.Rand
+	target *prog.Target
+
+	choiceTable *choiceTableProxy
+
+	nextExec     *priorityQueue[*Request]
+	runningExecs map[*Request]time.Time
+	jobIDs       map[job]int64
+	nextJobID    int64
+
+	queuedCandidates atomic.Uint64
+	// If the source of candidates runs out of them, we risk
+	// generating too many needCandidate requests (one for
+	// each Config.MinCandidates). We prevent this with candidatesRequested.
+	candidatesRequested atomic.Bool
+}
+
+func NewFuzzer(ctx context.Context, cfg *Config, rnd *rand.Rand,
+	target *prog.Target) *Fuzzer {
+	f := &Fuzzer{
+		Config: cfg,
+		Corpus: newCorpus(),
+
+		ctx:          ctx,
+		stats:        map[string]uint64{},
+		rnd:          rnd,
+		target:       target,
+		choiceTable:  &choiceTableProxy{},
+		nextExec:     makePriorityQueue[*Request](),
+		runningExecs: map[*Request]time.Time{},
+		jobIDs:       map[job]int64{},
+	}
+	go func() {
+		<-ctx.Done()
+		close(f.NewInputs)
+	}()
+	go f.leakDetector()
+	f.updateChoiceTable()
+	return f
+}
+
+type Config struct {
+	Logf           func(level int, msg string, args ...interface{})
+	Coverage       bool
+	FaultInjection bool
+	Comparisons    bool
+	Collide        bool
+	EnabledCalls   map[*prog.Syscall]bool
+	NoMutateCalls  map[int]bool
+	LeakChecking   bool
+	FetchRawCover  bool
+	// If the number of queued candidates is less than MinCandidates,
+	// NeedCandidates is triggered.
+	MinCandidates  uint
+	NeedCandidates chan struct{}
+	NewInputs      chan rpctype.Input
+}
+
+type Request struct {
+	Prog         *prog.Prog
+	NeedCover    bool
+	NeedRawCover bool
+	NeedSignal   bool
+	NeedHints    bool
+	// Fields that are only relevant within pkg/fuzzer.
+	flags   ProgTypes
+	stat    string
+	result  *Result
+	resultC chan *Result
+}
+
+type Result struct {
+	Info *ipc.ProgInfo
+	Stop bool
+}
+
+func (fuzzer *Fuzzer) Done(req *Request, res *Result) {
+	// Triage individual calls.
+	// We do it before unblocking the waiting threads because
+	// it may result it concurrent modification of req.Prog.
+	if req.NeedSignal && res.Info != nil {
+		for call, info := range res.Info.Calls {
+			fuzzer.triageProgCall(req.Prog, &info, call, req.flags)
+		}
+		fuzzer.triageProgCall(req.Prog, &res.Info.Extra, -1, req.flags)
+	}
+	// Unblock threads that wait for the result.
+	req.result = res
+	if req.resultC != nil {
+		req.resultC <- res
+	}
+	// Update stats.
+	fuzzer.mu.Lock()
+	fuzzer.stats[req.stat]++
+	delete(fuzzer.runningExecs, req)
+	fuzzer.mu.Unlock()
+}
+
+func (fuzzer *Fuzzer) triageProgCall(p *prog.Prog, info *ipc.CallInfo, call int,
+	flags ProgTypes) {
+	prio := signalPrio(p, info, call)
+	if !fuzzer.Corpus.AddRawMaxSignal(info.Signal, prio) {
+		return
+	}
+	fuzzer.Logf(2, "found new signal in call %d in %s\n",
```

Comment:
Log message shouldn't end with \n.
We have a check for this in tools/syz-linter, but it's checking only log.Logf. What happens if you change syz-linter to check all Logf calls? There shouldn't be Logf method names that are not doing logging.

Reply:
> What happens if you change syz-linter to check all Logf calls? There shouldn't be Logf method names that are not doing logging.

We specialize by specific `Logf` calls because they are not all the same:

https://github.com/google/syzkaller/blob/352ab9047be19ed1d8367b9113b7bde280c90124/tools/syz-linter/linter.go#L292-L296

Reply:
We can do:

```
--- a/tools/syz-linter/linter.go
+++ b/tools/syz-linter/linter.go
@@ -252,7 +252,7 @@ func (pass *Pass) checkFlagDefinition(n *ast.CallExpr) {
 // checkLogErrorFormat warns about log/error messages starting with capital letter or ending with a period.
 func (pass *Pass) checkLogErrorFormat(n *ast.CallExpr) {
        arg, newLine, sure := pass.logFormatArg(n)
-       if arg == -1 {
+       if arg == -1 || len(n.Args) <= arg {
                return
        }
        val, ok := stringLit(n.Args[arg])
@@ -302,6 +302,9 @@ func (pass *Pass) logFormatArg(n *ast.CallExpr) (arg int, newLine, sure bool) {
                }
                return 1, true, true
        }
+       if fun.Sel.String() == "Logf" {
+               return 1, false, true
+       }
        return -1, false, false
 }
```

It's clean now.

Reply:
Thanks! I've added it as a separate commit.